### DISCUSSION TITLE
Optimized Vector Map Pipeline

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#332] [Cross-project feature proposals]
## An Optimized MapLibre Display Pipeline

Here's some functionality we'd like to add to MapLibre.  This is one proposal of several for Native and if you'd like a little context and our strategy for the Web, [there's an overview](https://www.wetdogweather.com/blog/maplibre-proposals).

This proposal is a little different in that it reaches beyond the MapLibre toolkit to suggest changes to the way style sheets and vector tiles are processed with the goal of making for a leaner, faster, simpler version of the MapLibre Native toolkit.  Web could follow after, using the same data formats and strategy.

### Making MapLibre Native (or Web) Smaller

Toolkit size is a huge issue for developers big and small.  Download size measurably impacts adoption on native and generally annoys users on the web.  But cutting down the toolkit size is very difficult.

The problem is functionality and flexibility.  On functionality, there is a core set that most developers use, sure, but they also dip into random parts of the rest.  To draw a circle around the most commonly used parts is to make everyone a little bit unhappy.

Style sheets are amazingly flexible, as are vector tiles.  Excessively so for how they're being used in the vast majority of cases.  Moreover there's a conversion from raw data to what rendering engines expect that's being done on every single device, every time.

All of that requires more code in the toolkit.  More complexity, more flexibility, more size.

### A Leaflet Shaped Hole in our Hearts

Is the community clambering for a teeny, efficient Leaflet-like toolkit for Mobile and maybe Web? One that does vector tiles and styles and such?  Sort of feels like they are.

There are a lot of reasons that isn't really practical.  The style sheet spec is big and complex. Conformance is hard. Performance is too. *'Be reasonable'* you might say. 

I've been in tech for a long time now and I don't count on reasonableness. Open source developers don't do a careful survey and analysis before they start writing code.  They just kind of do it and hope they can find the resources to keep going.  Better to harness that urge than to fight it.

### Change the Problem to Fix the Problem

Trying to make the existing toolkits smaller is... worthwhile, but limited. Let's take a step back and look at the problem.

Developers want to display vector or image tiles with a particular style.  What takes up all the space in the toolkit:
- Style sheets are flexible.
    * The parser has to accept a lot of different concepts
    * And then hold those concepts in flexible data structures
    * Then help link those concepts to data
- Vector tiles are also flexible.
    * There's the format itself.
    * Then there's the tesselation.
    * Matching the features up to the style.
    * Converting vector geometry to renderable geometry.
- Platform interfaces are broad, as they should be. But they take up a surprising amount of space.

What if you decided you're just going to get vector and image tiles into users' eyeballs with the minimum of overhead.  What would that look like?

### Compiled Style Sheets

Style sheets are hugely flexible.  You don't really know what's in them until you've parsed them.  They're more like source code than binaries. They probably should be more like binaries.

What about a binary format for compiled style sheets? Fast to load, easy to predict what's in there.  

Really, it's the idea that you're going to be manipulating style sheets before they get to a client that's useful.  I know people are already doing this to a certain extent, so why not hop on that train and formalize it?

Think of a simplified style sheet format with all the various complexity reduced and expressions flattened. But to do that you'll need process vector tiles.  'Link' them, if you will.

### Compiled Vector Tiles

Vector tiles are already binary in form, but are they in concept really? I'd say no. They require linking with style sheets, a conversion to something renderable, not to mention tesselation.

For a given feature you have to ask the question "What style is this?"  And that question may break down into "It's green if it's above this zoom level and then red below that zoom level but only if this value is set to park".  Personally I'd prefer if the answer was something like "4".

To accomplish this seemingly simple feat (not simple) you'd have to process each vector tile on its way out to the client and run that expression logic.  While you're at it, why not also:
- Group things that can be rendered together.
- Tesselate everything. Yes, it'll be bigger, but honestly this is just irritating.
- Turn every style related question into a simple lookup.
- Toss out anything the user won't see.

Again, there are people doing some of these things already.  [Here are](https://github.com/mactrem/cov-tiles) some great ideas about format, for example.

### Vector Map Pipeline

If you pull the style sheet out into a compilable form and you commit to manipulating vector tiles on their way out the door, you create the possibility of some really interesting optimization.

This starts to look like a more complex vector map pipeline, like the ones we suspect the big companies have.  *Okay, I'm sure some of us know, but can't really talk about it.*  In any case, it's a chance to formalize a bit of what they do and add to it for the good of everyone.

I'm suggesting you develop and support those new formats by just adding them to the existing MapLibre to start. First Native, since that's where the big size and latency obsession is, then for the web.  A more optimized approach would have startup and latency benefits for the existing toolkits.

### Teeny Tiny Toolkits

By reducing the things a vector map display has to do, we reduce the complexity of the display software. Perhaps to the point where developers can just target those reduced formats.

So maybe you don't even have to write the Teeny Tiny Toolkit.  Maybe someone else will do it for you, according to your specs. I'd argue that's not necessarily the outcome you want.  Better to own it.  But if someone's going to do it, at least it involves your toolchain.

If you do want to own it, you have a head start.  MapLibre would already support the formats if you developed them. Just borrow your own code, throw out everything else and see how small you can make things.

### Discussion

You could put numbers to vector tile size and startup time and such, but I think this more a question of **vibes**.  Is the geospatial community going to will a new toolkit into being?  I think they might.

This approach would get you some new tools, support for an optimized version of the various formats and hedge against getting replaced down the road.

**Comment by wipfli** [comment]
Can users easily migrate existing styles to this new system?

**Reply by sjg-wdw** [reply]
First, they wouldn't need to.  The full MapLibre would continue supporting pure, uncompiled style sheets and unmodified vector tiles.

If they wanted to gain the advantages, then it would be a compile step.  Just a program to take a style sheet in and write out the new format.

Some of those advantages would point to restructuring vector tiles.  Ways to simplify style/feature mapping, tessellation, that sort of thing.  Then you've got a new vector tile format to deal with.  

You'd want tools to convert from the old vector tile format to a new one, with optional support for style-based features.  I'd also suggest a very simple service to convert vector tiles on the fly.  Something you could put in the chain between client, CDN, and tile source.

Each of these formats could list what they contain.  Not just the format version, but a survey of the contents.  Does this style contain expressions?  What kind of expressions?  Does the vector tile contain pre-tessellated data?  How about generic string/value keys?

By the time you get to the Teeny Tiny Toolkits, they can support a very stripped down version of the original flexibility and check within the style and vector formats to see what's there.  Ideally solving the "where did my roads go?" problem.
